packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,406 @@
+		 *
+		 * @param int $id The ID of the entity.
+		 * @param int $session_id The current session ID.
+		 * @param int $entity_type The type of the entity, comment, etc.
+		 * @param string $entity_id The ID of the entity before the import.
+		 * @param string $mapped_id The mapped ID of the entity after the import.
+		 * @param string $parent_id The parent ID of the entity.
+		 * @param string $additional_id The additional ID of the entity. Used for comments and terms. Comments have a comment_parent, and the post.
+		 * @param int $byte_offset The byte offset of the entity inside the WXR file. Not used now.
+		 * @param int $sort_order The sort order of the entity. Not used now.
+		 */
+		$sql = $wpdb->prepare(
+			'CREATE TABLE IF NOT EXISTS %i (
+				id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+				session_id bigint(20) unsigned,
+				entity_type tinyint(1) NOT NULL,
+				entity_id text NOT NULL,
+				mapped_id text DEFAULT NULL,
+				parent_id text DEFAULT NULL,
+				additional_id text DEFAULT NULL,
+				byte_offset bigint(20) unsigned NOT NULL,
+				sort_order int DEFAULT 1,
+				PRIMARY KEY  (id),
+				KEY session_id (session_id),
+				KEY entity_id (entity_id(%d)),
+				KEY parent_id (parent_id(%d)),
+				KEY byte_offset (byte_offset)
+			) ' . $wpdb->get_charset_collate(),
+			self::get_table_name(),
+			$max_index_length,
+			$max_index_length
+		);
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		dbDelta( $sql );
+	}
+
+	/**
+	 * Run by register_deactivation_hook. It drops the table and deletes the
+	 * option.
+	 */
+	public static function deactivate() {
+		global $wpdb;
+		$table_name = self::get_table_name();
+
+		// Drop the table.
+		$wpdb->query( $wpdb->prepare( 'DROP TABLE IF EXISTS %s', $table_name ) );
+
+		// Delete the option.
+		delete_option( self::OPTION_NAME );
+	}
+
+	/**
+	 * Reset the class.
+	 */
+	public function reset() {
+		$this->set_session( null );
+	}
+
+	/**
+	 * Set the current session ID.
+	 *
+	 * @param int|null $session_id The session ID.
+	 */
+	public function set_session( $session_id ) {
+		$this->current_session = $session_id;
+	}
+
+	/**
+	 * Delete all rows for a given session ID.
+	 *
+	 * @param int $session_id The session ID to delete rows for.
+	 * @return int|false The number of rows deleted, or false on error.
+	 */
+	public function delete_session( $session_id = null ) {
+		global $wpdb;
+
+		return $wpdb->delete(
+			self::get_table_name(),
+			array( 'session_id' => $session_id ?? $this->current_session ),
+			array( '%d' )
+		);
+	}
+
+	/**
+	 * Map an entity to the index. If $id is provided, it will be used to map the entity.
+	 *
+	 * @param string   $entity_type The type of the entity.
+	 * @param array    $data The data to map.
+	 * @param int|null $id The ID of the entity.
+	 * @param int|null $additional_id The additional ID of the entity.
+	 */
+	public function map_entity( $entity_type, $data, $id = null, $additional_id = null ) {
```

Comment:
Let's use a method name and docstring that thoroughly explain what this method does, even if it takes a few paragraphs of text. What's the index? Where does it live? What is its layout? What is it used for? Why an `$id` is needed? What is the relevance of `$additional_id`? What if one of them is missing? Can this method fail, and how do we know?